clients/src/main/java/org/apache/kafka/clients/consumer/internals/AbstractCoordinator.java

```diff
@@ -352,6 +332,44 @@ else if (!future.isRetriable())
         }
     }
 
+    private synchronized void resetJoinGroupFuture() {
+        this.joinFuture = null;
```

Comment:
I was concerned that the join group response handler could get invoked from the background thread (e.g. if the user calls wakeup() before the join finishes), but I guess that should be impossible since we disable the heartbeat thread prior to beginning the join and we don't reenable it until it returns. Does that sound right?